### PR TITLE
Enable CUDA eager and inductor

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request enables CUDA eager and inductor for the blog branch.

1. It introduces a env called: USE_CUDA to bypass any xla init which somehow could slows down the CUDA speed.
2. It then utilizes that USE_CUDA env to distinguish scenarios where torch.dist and inductor need to be used.
3. It also introduces example_cuda.py as the driver to serve the model given torch.dist initialization is vastly different from XLA.
4. Finally, it adds a tweak for the T5 tokenizer as it doesn't have `bos` which will error out the embedding in CUDA.

Test Plan:
USE_CUDA=1 python example_cuda.py --tokenizer_path ../llama/spiece.model --max_seq_len 256 --max_batch_size 1 --temperature 0.8 --dim 4096 --n_heads 32 --n_layers 32 --mp True
PJRT_DEVICE=GPU GPU_NUM_DEVICES=4 python example_xla.py --tokenizer_path ../llama/spiece.model --max_seq_len 256 --max_batch_size 1 --temperature 0.8 --dim 4096 --n_heads 32 --n_layers 32 --mp True

### CODE REVIEW

**Comment by alanwaketan** [comment]
Can I get a review?

**Comment by JackCaoG** [comment]
I will take a look today

File: example_cuda.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    if rank > 0:
+        sys.stdout = open(os.devnull, "w")
+
+    generator = load(ckpt_dir, tokenizer_path, rank,
+                     world_size, max_seq_len, max_batch_size,
+                     torch.device("cuda", rank), dim, n_layers, n_heads)
+
+    prompts = [
+        # For these prompts, the expected answer is the natural continuation of the prompt
+        "I believe the meaning of life is",
```
Comment (JackCaoG) [inline example_cuda.py:53]: do we need to update this part to also take variable length prompts? I guess this can be done as a follow up if we think it is worth the time.

File: example_xla.py [hunk 2]
```diff
@@ -7,14 +7,20 @@
...
 from pathlib import Path
 
 from llama import ModelArgs, Transformer, Tokenizer, LLaMA
 from llama.xla_model_parallel import get_model_parallel_rank, get_model_parallel_world_size
 
+import os
+
+USE_CUDA = os.environ.get('USE_CUDA', False)
+
+if not USE_CUDA:
```
Comment (JackCaoG) [inline example_xla.py:20]: can you add a comment here to explian why we have to do conditional import?

File: llama/generation.py [hunk 3]
```diff
@@ -5,35 +5,50 @@
...
-import torch_xla.core.xla_model as xm
 
 from llama.tokenizer import Tokenizer
 from llama.model import Transformer
 
+import os
+
+USE_CUDA = os.environ.get('USE_CUDA', False)
+
+if not USE_CUDA:
```
Comment (JackCaoG) [inline llama/generation.py:16]: ditto

File: llama/model.py [hunk 4]
```diff
@@ -194,23 +210,46 @@ def __init__(
...
-        self.w1 = ColumnParallelLinear(
-            dim, hidden_dim, bias=False, gather_output=False, init_method=init_method, world_size=world_size, rank=rank, groups=groups, quant=quant
-        )
-        self.w2 = RowParallelLinear(
-            hidden_dim, dim, bias=False, input_is_parallel=True, init_method=init_method, world_size=world_size, rank=rank, groups=groups, quant=quant
-        )
-        self.w3 = ColumnParallelLinear(
-            dim, hidden_dim, bias=False, gather_output=False, init_method=init_method, world_size=world_size, rank=rank, groups=groups, quant=quant
-        )
+        self.w1 = ColumnParallelLinear(dim,
```
Comment (JackCaoG) [inline llama/model.py:213]: did you use a different formater? A lot of code changes to python are just format changes.

**Review by JackCaoG — APPROVED** [review]
mostly LGTM, there are a lot of python format change which makes code review difficult.. Can you use this pr and run a sanity check on TPU? It touched a bunch of files, want to make sure it does not introduce any unwated regression(it shouldn't..)

**Review by JackCaoG — APPROVED** [review]
mostly LGTM, there are a lot of python format change which makes code review difficult.. Can you use this pr and run a sanity check on TPU? It touched a bunch of files, want to make sure it does not introduce any unwated regression(it shouldn't..)

File: example_cuda.py (hunk 1 above)
Comment (alanwaketan) [reply example_cuda.py:53]: This change only targets the blog branch not the stable branch for Meta.

**Comment by alanwaketan** [comment]
> mostly LGTM, there are a lot of python format change which makes code review difficult.. Can you use this pr and run a sanity check on TPU? It touched a bunch of files, want to make sure it does not introduce any unwated regression(it shouldn't..)

I did that already on both TPU and XLA:GPU. Things looks good to me. Thanks for the reminder and review.

File: llama/model.py (hunk 4 above)
Comment (alanwaketan) [reply llama/model.py:213]: I just use yapf...

File: llama/generation.py [hunk 5]
```diff
@@ -56,14 +73,17 @@ def generate(
         params = self.model.params
         assert bsz <= params.max_batch_size, (bsz, params.max_batch_size)
 
-        prompt_tokens = [self.tokenizer.encode(x, bos=True, eos=False) for x in prompts]
+        bos = not USE_CUDA  # The supplied t5 tokenizer doesn't have bos. CUDA will error out but xla won't.
```
Comment (Liyang90) [inline llama/generation.py:76]: Sorry for the late comment. Would this affect the users that try to use the LLaMA checkpoint and tokenizer?

Comment (alanwaketan) [reply llama/generation.py:76]: Good point. Yea, it would.  Is there a way to tell whether this is a T5 tokenizer or LLaMA? I can do a follow up.

Comment (Liyang90) [reply llama/generation.py:76]: Maybe a `try...except..` to catch and fall back to `False`?

Comment (alanwaketan) [reply llama/generation.py:76]: How? The error only manifest in the later embedding access code: https://github.com/pytorch-tpu/llama/blob/main/llama/model.py#L225. The CUDA kernel crashes...
